### ISSUE TITLE
For TextEdit/LineEdit, provide a checkbox in the editor to automatically select text on any focus/mouse entered events.

### ISSUE THREAD

**Issue by wyattbiker** [godotengine/godot-proposals#171]
**Describe the project you are working on:**
Any UI project

**Describe the problem or limitation you are having in your project:**
When a **LineEdit**/**TextEdit** control receives focus automatically select the text and deselect when focus is lost. Otherwise I need to create a connection for `focus_entered, focus_exited(), mouse_entered(), mouse_exited()` and call the functions `select_all(), deselect()` . It is a PITA.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
The enhancement would add a checkbox in the editor Line/TextEdit properties called **Select Text on Focus**, so that when an input control gets focus (_whether via mouse or tabbing or focus_grab())_  to automatically select the text.

In addition there should be an opposite checkbox **Deselect Text on Lost Focus**.

This is essential especially handy for, but not limited to),numeric values. Normally numeric values are not edited by cursoring over them but just overtyped.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**

Example tabbing/mousing Godot numeric values:
![image](https://user-images.githubusercontent.com/930478/67162517-b3162c80-f332-11e9-8d44-8a24fb51ff88.png)

Example Unity if you tab in/out or mouse/in out it highlights the columns. Many other applications

![image](https://user-images.githubusercontent.com/930478/67162091-ece53400-f32e-11e9-9394-93a9c5af3baa.png)

**Describe implementation detail for your proposal (in code), if possible:**
As stated above, if Select/Deselect Text is checked, when any of the inputs receive focus, to select the text, and the opposite when they lose focus.

SpinBox and RichText (although I think these have specialized purpose) would possibly fall into this :)

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
UIs will be using this often. Also plugins for the editor.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It should be core.

**Comment by Calinou** [comment]
As I said on IRC, I think this is better implemented as a default behavior in SpinBox, as it's intended to be used for numeric input. This way, we don't need to add yet another property to LineEdit and TextEdit.

**Comment by wyattbiker** [comment]
You are correct this can be implemented in code. Like many other things :). But SpinBox has very specific numerics only purpose. It also has range limits and increments. Does not auto-select either. And shows an extra double arrow control attached to it.  However not everything is strictly numeric entry. For example, ID's, ZipCodes, Phone Numbers, etc.  

User input validation is one of the areas where an application fails or succeeds. The more code a programmer has to do that is of the mundane kind, the more testing needs to be done catch items that are validated with a simple checkbox. 

I know this is a game engine, but I believe it could be used as an easy general purpose tool to create a mobile app. Why not?

**Comment by girng** [comment]
> I need to create a connection for `focus_entered, focus_exited(), mouse_entered(), mouse_exited()` and call the functions `select_all(), deselect()` . It is a PITA.

This is not a negative, it's the power of GDScript.

**Comment by bojidar-bg** [comment]
I guess you can just connect `focus_entered` directly to `select_all` and call it a day, no?

**Comment by wyattbiker** [comment]
@bojidar-bg  Not to beat up on this horse (there are more pressing issues with Godot), but you need to connect to mouse_entered as well and then deselect in the focus_exited and mouse_exited() connections. Otherwise the select remains highlighted if you are just tabbing thru.

**Comment by bojidar-bg** [comment]
@wyattbiker I don't think you should connect to `mouse_entered` and `mouse_exited`. Either way, all these connections should be doable without a script using the advanced connection GUI.

**Comment by Ramh5** [comment]
@bojidar-bg  if you only connect `focus_entered` and exited `select_all()` does not work when gaining focus with the mouse even though the focus signal does get sent because I guess the mouse event deselects on on clicking after the focus is gained.

**Comment by RebelliousX** [comment]
I can confirm this, select_all() only works with RIGHT click in _gui_input() event.
Setting mouse filter to pass or stop doesn't affect anything, no other node covering LineEdit.

**Comment by kitbdev** [comment]
- `select_all_on_focus` was added for LineEdit in https://github.com/godotengine/godot/pull/67489
- `deselect_on_focus_loss_enabled` was added for TextEdit and LineEdit in https://github.com/godotengine/godot/pull/54166

So there is only `select_all_on_focus` that has not yet been added for TextEdit.
